src/XMonad/StackSet.hs

```diff
@@ -517,6 +520,24 @@ delete' w s = s { current = removeFromScreen        (current s)
     where removeFromWorkspace ws = ws { stack = stack ws >>= filter (/=w) }
           removeFromScreen scr   = scr { workspace = removeFromWorkspace (workspace scr) }
 
+-- | Map over the 'Workspace's of a 'StackSet'.
+mapWorkspaces
+  :: (Workspace i l a        -> Workspace i' l' a)
+  ->  StackSet  i l a sid sd -> StackSet  i' l' a sid sd
+mapWorkspaces f = runIdentity . traverseWorkspaces (Identity . f)
+
+-- | 'traverse' the 'Workspace's of a 'StackSet'.
+traverseWorkspaces
+  :: Applicative f
+  => (Workspace i l a        -> f (Workspace i' l' a))
+  ->  StackSet  i l a sid sd -> f (StackSet  i' l' a sid sd)
+traverseWorkspaces f s = StackSet
+  <$>          onScreen (current  s)
+  <*> traverse onScreen (visible  s)
+  <*> traverse f        (hidden   s)
+  <*> pure              (floating s)
+  where onScreen scr = f (workspace scr) <&> \w -> scr{ workspace = w }
+
```

Comment:
Maybe we should reconsider https://github.com/xmonad/xmonad/pull/357 after all...